packages/arb-validator/checkpointing/indexedCheckpointer.go

```diff
@@ -0,0 +1,477 @@
+	bounds, err := cp.getHeightBounds()
+	if err != nil {
+		return err
+	}
+	if bounds == nil {
+		bounds = &ckpHeightBounds{
+			lo: height.Clone(),
+			hi: height.Clone(),
+		}
+		return cp.setHeightBounds(bounds)
+	} else if bounds.hi.Cmp(height) < 0 {
+		bounds.hi = height.Clone()
+		return cp.setHeightBounds(bounds)
+	} else {
+		return nil
+	}
+}
+
+func (cp *IndexedCheckpointer) getIdsAtHeight(height *common.TimeBlocks) ([]*common.BlockId, error) {
+	key := append([]byte{1}, height.AsInt().Bytes()...)
+	valBytes := cp.db.GetData(key)
+	if valBytes == nil {
+		return nil, nil
+	}
+	idBufList := &BlockIdBufList{}
+	if err := proto.Unmarshal(valBytes, idBufList); err != nil {
+		return nil, err
+	}
+	ret := []*common.BlockId{}
+	for _, idBuf := range idBufList.Bufs {
+		ret = append(ret, idBuf.Unmarshal())
+	}
+	return ret, nil
+}
+
+func (cp *IndexedCheckpointer) setIdsAtHeight(height *common.TimeBlocks, ids []*common.BlockId) error {
+	key := append([]byte{1}, height.AsInt().Bytes()...)
+	idBufs := []*common.BlockIdBuf{}
+	for _, id := range ids {
+		idBufs = append(idBufs, id.MarshalToBuf())
+	}
+	idBufList := &BlockIdBufList{
+		Bufs: idBufs,
+	}
+	valBytes, err := proto.Marshal(idBufList)
+	if err != nil {
+		return err
+	}
+	ok := cp.db.SaveData(key, valBytes)
+	if !ok {
+		return errors.New("db write error in checkpointer.setIdsAtHeight")
+	}
+	return nil
+}
+
+func (cp *IndexedCheckpointer) deleteIdsAtHeight(height *common.TimeBlocks) error {
+	key := append([]byte{1}, height.AsInt().Bytes()...)
+	if ok := cp.db.DeleteData(key); !ok {
+		return errors.New("Checkpointer failed to delete idsAtHeight")
+	}
+	return nil
+}
+
+func (cp *IndexedCheckpointer) recordIdAsCheckpointed(newId *common.BlockId) error {
+	ids, err := cp.getIdsAtHeight(newId.Height)
+	if err != nil {
+		return err
+	}
+	ids = append(ids, newId)
+	return cp.setIdsAtHeight(newId.Height, ids)
+}
+
+func (cp *IndexedCheckpointer) GetInitialMachine() (machine.Machine, error) {
+	cp.Lock()
+	defer cp.Unlock()
+
+	return cp.db.GetInitialMachine()
+}
+
+type writableCheckpoint struct {
+	blockId  *common.BlockId
+	contents []byte
+	ckpCtx   CheckpointContext
+}
+
+func (cp *IndexedCheckpointer) AsyncSaveCheckpoint(
+	blockId *common.BlockId,
+	contents []byte,
+	cpCtx CheckpointContext,
+	closeWhenDone chan struct{},
+) {
+	cp.Lock()
+	defer cp.Unlock()
+
+	if closeWhenDone != nil {
+		cp.chansToClose = append(cp.chansToClose, closeWhenDone)
+	}
+	cp.nextCheckpointToWrite = &writableCheckpoint{
+		blockId:  blockId,
+		contents: contents,
+		ckpCtx:   cpCtx,
+	}
+}
+
+func (_ *IndexedCheckpointer) makeContentsKey(id *common.BlockId) []byte {
+	bidBuf := id.MarshalToBuf()
+	bytesBuf, err := proto.Marshal(bidBuf)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return append([]byte{2}, bytesBuf...)
+}
+
+func (cp *IndexedCheckpointer) RestoreLatestState(
+	ctx context.Context,
+	clnt arbbridge.ArbClient,
+	_ common.Address,
+	_ bool,
```

Comment:
Are these parameters used anywhere? If not they should be removed.

Reply:
Eliminated these params, here and in the interface that this is implementing.  (No implementation of the interface used them.)